r3bsource/pspx/R3BS515PspxReader.cxx

```diff
@@ -0,0 +1,265 @@
+ */
+Bool_t R3BS515PspxReader::Read()
+{
+    EXT_STR_h101_PSP_onion* data = reinterpret_cast<EXT_STR_h101_PSP_onion*>(fData);
+
+#if 0
+    // this is the data structure we have to read:
+    struct {
+      struct {
+        struct {
+          uint32_t E;
+          uint32_t EI[32 /* E */];
+          uint32_t Ev[32 /* E */];
+        } S[2];
+      } F[2];
+    } PSPX[N];
+
+    F = Face (front/back of detector)
+    S = Side (end of strip)
+#endif
+
+    if (fShift > 0)
+    {
+        // skip TPAT = 0 events :
+        if (header->GetTpat() == 0)
+        {
+            return kTRUE;
+        }
+        // store number = fShift events into buffer
+        if (fBuffer_i < fShift)
+        {
+            *fEventBuffer[fBuffer_i] = *data;
+            fBuffer_i++;
+            if (fBuffer_i == fShift)
+            {
+                fBuffer_i = fShift + 1;
+            }
+        }
+        // now instead of reading current event from "data" directly  we read from fEventBuffer
+        // reading the event:
+        // loop over all detectors
+        if (fBuffer_i == fShift)
+        {
+            // read from Event_buffer[0]
+            for (Int_t d = 0; d < LENGTH(fEventBuffer[0]->PSPX); d++)
+            {
+                // loop over faces
+                for (Int_t f = 0; f < 2; f++)
+                {
+                    std::vector<R3BPspxMappedData*> datas(LENGTH(fEventBuffer[0]->PSPX[0].F[0].S[0].EI));
+                    // loop over strip sides
+                    for (Int_t s = 0; s < 2; s++)
+                    {
+                        auto const& dfs = fEventBuffer[0]->PSPX[d].F[f].S[s];
+                        uint32_t numChannels = dfs.E;
+                        // loop over channels
+                        for (Int_t i = 0; i < numChannels; i++)
+                        {
+                            int32_t strip = dfs.EI[i]; // counting from 1 to max number of channels for an detector
+                            int32_t energy = dfs.Ev[i];
+                            if (energy < 0)
+                                energy =
+                                    -1. *
+                                    energy; // make sure energy values are positive. Necessary for compatibilty with
+                                            // GSI Febex firmware
+
+                            // if (!datas[strip-1]) datas[strip-1] = new
+                            // ((*fMappedItems[d])[fMappedItems[d]->GetEntriesFast()])
+                            // R3BPspxMappedData(f+1,strip);
+                            if (!datas[strip - 1])
+                                datas[strip - 1] =
+                                    new ((*fMappedItems[2 * d + f])[fMappedItems[2 * d + f]->GetEntriesFast()])
+                                        R3BPspxMappedData();
+                            // assert(-1 == datas[strip-1]->GetEnergy(s));
+                            datas[strip - 1]->SetValue(s, strip, energy);
```

Comment:
Ok, this comes from the original psp reader. I'm really amazed how this was handled.

In the beginning, I'm already wondering why there is a vector of TCloneArray. Isn't one TCloneArray enough? And then I came down here, still wondering it. Why can't we just have one TCloneArray? Then it would simply be like:
```cpp
auto mapdata = new (fMapdata->GetEntriesFast()) R3BPspxMappedData(s, strip, energy);
```
Do you know who wrote the original psp reader?

Reply:
--